hs-bindgen/src/HsBindgen/Hs/AST.hs

```diff
@@ -0,0 +1,85 @@
+-- | Haskell AST
+--
+-- This is a bespoke AST, specifically targetted at the hs-bindgen use case. We
+-- do not attempt to cover all of Haskell, and we provide a more strongly typed
+-- AST. For example, rather than providing an abstraction for "general class
+-- instance", we provide abstractions for, say, a 'Storable' instance. This
+-- makes the AST more strongly typed, whilst at the same time providing the
+-- various backends (TH, preprocessor, ..) with the information that they need.
+--
+-- This AST does not specify variable names (for variables where the name does
+-- not matter): we leave it up to specific backends to generate fresh names.
+module HsBindgen.Hs.AST where
+
+{-------------------------------------------------------------------------------
+  Declarations
+-------------------------------------------------------------------------------}
+
+-- | Top-level declaration
+data Decl f =
+    DeclInstance (InstanceDecl f)
+
+-- | Class instance declaration
+data InstanceDecl f =
+    InstanceStorable (StorableInstance f)
+
+{-------------------------------------------------------------------------------
+  Expressions
+-------------------------------------------------------------------------------}
+
+data Expr f =
+    ExprVar (f Bound)
+
+{-------------------------------------------------------------------------------
+  Structs
+-------------------------------------------------------------------------------}
+
+-- | Construct value of a struct
+data IntroStruct f = IntroStruct Free [Expr f]
+
+-- | Pattern match on a struct
+data ElimStruct f a = ElimStruct Free ([f Bound] -> a)
```

Comment:
I'm not sure I like HOAS. It's not `Show`able or generally inspectable. Also here, the arity is not encoded.

Reply:
This is not HOAS, it's PHOAS, and therefore it is in fact showable. And yes, arity isn't encoded; I'm not sure exactly how precise we want to be in this AST. I nearly have a TH translation for this done (along with some cleanup). We can discuss details then.

Reply:
derive `Show`. I'd like to write tests comparing the ASTs (i.e. golden tests) before backends. I only see that possible using a "mock" backend.

Reply:
Yes, I agree that we should have `Show` instances.

Reply:
I think actually we should just have `ToExpr` instances in the main library, rather than declaring them as orphans.

Reply:
For users of the `hs-bindgen` as library (if they use TH generation), each extra dependency is an extra dependency.

Reply:
Yeah, that's a fair point.

Reply:
Especially as `tree-diff` is not particularly lean, it pulls in `aeson` e.g. because it's a batteries-included library. Having batteries is nice in testing (to avoid dzillion of packages, c.f. `data-default`), but not so for dependency footprint.

Reply:
Perhaps we can actually flip it around: use a first order representation but provide a HOAS API. I'll try that.

Reply:
Although I find the translation to TH in this PHOAS form quite pleasant.

Reply:
On second thoughts, I'll leave it like this for now (but will provide a `Show` instance for closed forms).